### ISSUE TITLE
ffmpeg-mux no longer silently ignoring errors

### ISSUE THREAD

**Issue by drbenvincent** [obsproject/obs-studio#3588]
## Platform
Operating system and version: MacOS 10.15.6
OBS Studio version: 26.0.2

## Current Behavior
Upon stating a recording, error window appears with "An encoder error occurred while recording". This stops recording, leaving a file with just a few recorded frames in it. This started immediately after upgrading to 26.0.2.

## Steps to Reproduce
1. Open OBS
2. Set up minimal scene with audio/visual input
3. Record
4. Error

## Additional information
Going into Settings > Output in an attempt to check/change settings, shows that all the options are greyed out. Nothing can be changed other than the bitrate, see image
![Screenshot 2020-10-09 at 13 17 00](https://user-images.githubusercontent.com/6765047/95581928-bd44a580-0a31-11eb-940e-5b69dd432469.png)
Going into Settings > Video gives an intriguing message of "Video output is currently active. Please turn off any outputs to change video settings". I'm not precisely sure what this means or how to turn off video output... Nothing had changed in my workflow or setup before the bug started.
![Screenshot 2020-10-09 at 13 17 47](https://user-images.githubusercontent.com/6765047/95582017-e6653600-0a31-11eb-98e1-005ed7c0ef20.png)
Attempted to downgrade, but noted that executables are only available for the latest version.

EDIT: Log file now attached
[2020-10-09 12-25-20.txt](https://github.com/obsproject/obs-studio/files/5354829/2020-10-09.12-25-20.txt)

**Comment by Scrxtchy** [comment]
As per your log file, the active output is the NDI broadcast
```
12:25:21.349: [obs-ndi] starting NDI main output with name 'OBS'
12:25:21.350: [obs-ndi] 'OBS': ndi output started
```
Otherwise this is a duplicate of #3572
> Please use the discord or the forums for support.

**Comment by drbenvincent** [comment]
Narrowed down the problem to Apply VT H264 Software Encoding as the encoder that causes the error. happy to use the forums, but what's the boundary between a software bug and support? It worked in the previous version, so felt like it was a bug.

**Comment by PatTheMav** [comment]
@drbenvincent Could you please check if the problem persists when changing to a different output folder? Also change the output container to `mkv` and try recording with that?

**Comment by pkviet** [comment]
can you try recording in flv? mp4 is never a wise choice actually as spelt out in the orange warning

**Comment by pkviet** [comment]
and also give us your hardware specs (macbook, imac, version ...)

**Comment by drbenvincent** [comment]
Hardware encoding + different folder + mp4 (as before)leads to same error.
Hardware encoding + different folder + mkv leads to same error.

Recording with x268 works fine. Recording with software H264 works, but seems to be terrible despite the same bitrate (maybe that's irrelevant).

Happy to give more info

**Comment by pkviet** [comment]
in 25.0.8 it worked fine ? or only on 26.0.2 ?

**Comment by drbenvincent** [comment]
Yes, it's been working fine - I've been recording for about a month so I believe that covers 25.0.8 and 26.0. Opened up to record today, did the autoupdate and all of a sudden the error started. No change in settings, OS update, hardware, etc.

**Comment by PatTheMav** [comment]
You can still download 26.0 from our releases page here on Github - can you just download and run OBS from the disk image (it should pick up your current setup just fine) and try with that version?

**Comment by drbenvincent** [comment]
> You can still download 26.0 from our releases page here on Github - can you just download and run OBS from the disk image (it should pick up your current setup just fine) and try with that version?

I tried that first of all, but it looks like it's just the source code available? https://github.com/obsproject/obs-studio/releases/tag/26.0.0  Sorry if I'm not looking in the right place.

**Comment by PatTheMav** [comment]
Huh, try the build from our CI: https://github.com/obsproject/obs-studio/actions/runs/276555583

The one ending with `-rel` in the filename is the notarised release version.

**Comment by eric** [comment]
Due to #3460 ffmpeg errors are now being handled and cause ffmpeg-mux to be aborted instead of silently ignoring the errors. 

There may be some transient error cases that need to be handled in ffmpeg-mux that are not handled.

**Comment by PatTheMav** [comment]
@Eric The errors reported in the logs seem to be emitted by the `write_packet` function when the return value of `ret = os_process_pipe_write(stream->pipe, packet->data, packet->size);` doesn't equal the packet size (either because less bytes were written or an error occurred and thus `0` is returned).

I haven't dug into ffmpeg-mux enough to rule this out, but that doesn't look like an area that your PR touched?

**Comment by RytoEX** [comment]
@drbenvincent 
Could you please try again with a recent macOS build?
https://github.com/obsproject/obs-studio/actions/runs/310675931

**Comment by eric** [comment]
@PatTheMav The thing that changed is that previously if there was a problem in the ffmpeg encoder in the mux process, it would silently ignore the error and continue. Now, with the change, it will see the error and exit the process. That could cause the `os_process_pipe_write for packet data failed` message. On Windows, the pipe process captures `stderr` (on POSIX you will just see the ffmpeg logs in the log).

So, there are a few possibilities:

1. There was a transient error that happens when calling `av_interleaved_write_frame()` in ffmpeg-mux that does not prevent the process from otherwise working that is now causing  ffmpeg-mux to exit and print that message
2. Something is for some reason isn't being printed in the logs by `signal_failure()`
3. `stderr` isn't being read from `ffmpeg-mux` until the process exits and it could be causing issues with the buffer of the child process getting full

**Comment by Fenrirthviti** [comment]
Several similar reports, I've updated the title of this PR and going to close the others and direct them here for further discussion.

**Comment by yedi-yanez** [comment]
Try changing the "Common FPS values" in settings -> video section to something different to 29.97 or 59.94. Thats work form me. No error while recording using "use stream coder" and the Apple VT h264 Hardware Encoder.

**Comment by jp9000** [comment]
I've reverted the line causing problems for now in https://github.com/obsproject/obs-studio/commit/ffe4c8552bb25a2e4ac8b88e60b6f3ac1784a7f8.

**Comment by RytoEX** [comment]
> 
> 
> I've reverted the line causing problems for now in [ffe4c85](https://github.com/obsproject/obs-studio/commit/ffe4c8552bb25a2e4ac8b88e60b6f3ac1784a7f8).

The revert above has been reverted by https://github.com/obsproject/obs-studio/commit/17b38735786f019cea64b03ed0659fcefa16ebaf with additional fixes.  Can those affected by this issue please try the current builds?  @drbenvincent 

https://github.com/obsproject/obs-studio/actions/runs/371349643

**Comment by RytoEX** [comment]
The fixes I mentioned above are also included in OBS Studio 26.1 RC1.  If someone previously affected could test RC1 and report back, that'd be great.  Thanks!
@drbenvincent

**Comment by cjhlambert** [comment]
I'm also having this issue. But I'm not tech savvy at all. I don't understand what anyone here is saying. And I can't fix it.

**Comment by kerosenesoundsystem** [comment]
> I'm also having this issue. But I'm not tech savvy at all. I don't understand what anyone here is saying. And I can't fix it.

i tried @yedi-yanez solution and it worked for me. go to settings -> video, and then in common FPS values change it to 30, i tried several values but only 30fps works for me. and then you can record using Apple Hardware encoder.

**Comment by dakhar48** [comment]
I am an OBS newbie, so apologize up front if that shows through. I had the same problem. Finally figured out that if VirtualCam is active in OBS, my video options in settings are all greyed out. Turning off VirtualCam resolves that for me. Hope this helps someone else.

**Comment by Fenrirthviti** [comment]
That is entirely unrelated to this issue.

**Comment by davegrodzki** [comment]
> The fixes I mentioned above are also included in OBS Studio 26.1 RC1. If someone previously affected could test RC1 and report back, that'd be great. Thanks!
> @drbenvincent

Hello... This is still preventing me from recording to anything except mkv in 26.1 final release. Is there anything that can be done to restore ability to record to MP4 or MOV with the Apple Hardware Encoder?

**Comment by Fenrirthviti** [comment]
These fixes should have made it in to the final release, can you please provide a log file after a record failure has occurred?

**Comment by davegrodzki** [comment]
See attached! Thanks so much!

[2020-12-23 13-03-11.txt](https://github.com/obsproject/obs-studio/files/5736730/2020-12-23.13-03-11.txt)

**Comment by Fenrirthviti** [comment]
Looks like you're using custom FFmpeg output mode, which unfortunately is a "use at your own risk" configuration in OBS, and doesn't have anything to do with this issue. I'd recommend stopping by the support channels in Discord to see if someone can take a closer look, but as mentioned we typically are hands off with that mode.

**Comment by PatTheMav** [comment]
While we're at it:

@drbenvincent Please confirm if this issue persists with 26.1 or has been fixed.

**Comment by davegrodzki** [comment]
Thanks so much 

> Looks like you're using custom FFmpeg output mode, which unfortunately is a "use at your own risk" configuration in OBS, and doesn't have anything to do with this issue. I'd recommend stopping by the support channels in Discord to see if someone can take a closer look, but as mentioned we typically are hands off with that mode.

Thanks so much for this suggestion sir. If I select Settings > Output > Recording > Type = Standard (NOT Custom Output FFmpeg), I can actually record to MOV/MP4 at 30/60fps, but not 29.97 or 59.94. I also cannot record to MP4/MOV if I check "Use B Frames" on the Output > Streaming (all of these things result in the same encoder error).

I am certainly not an advanced user, but am wondering if this helps for others.

13:51:23.504: [VideoToolbox streaming_h264: 'h264']: session created with hardware encoding
13:51:23.509: [VideoToolbox streaming_h264: 'h264']: settings:
13:51:23.509: 	vt_encoder_id          com.apple.videotoolbox.videoencoder.h264.gva
13:51:23.509: 	bitrate:               25000 (kbps)
13:51:23.509: 	fps_num:               60
13:51:23.509: 	fps_den:               1
13:51:23.509: 	width:                 1920
13:51:23.509: 	height:                1080
13:51:23.509: 	keyint:                2 (s)
13:51:23.509: 	limit_bitrate:         off
13:51:23.509: 	rc_max_bitrate:        2500 (kbps)
13:51:23.509: 	rc_max_bitrate_window: 1.500000 (s)
13:51:23.509: 	hw_enc:                on
13:51:23.509: 	profile:               high
13:51:23.509: 
13:51:23.513: [CoreAudio AAC: 'Main Mix']: settings:
13:51:23.513: 	mode:          AAC
13:51:23.513: 	bitrate:       320
13:51:23.513: 	sample rate:   48000
13:51:23.513: 	cbr:           on
13:51:23.513: 	output buffer: 1536
13:51:23.516: ==== Recording Start ===============================================
13:51:23.516: [ffmpeg muxer: 'adv_file_output'] Writing file '/Users/davegrodzki/Movies/2020-12-23 13-51-23.mp4'...
13:51:23.761: [ffmpeg muxer: 'adv_file_output'] os_process_pipe_write for packet data failed
13:51:23.761: [ffmpeg muxer: 'adv_file_output'] Output of file '/Users/davegrodzki/Movies/2020-12-23 13-51-23.mp4' stopped
13:51:23.761: Output 'adv_file_output': stopping
13:51:23.761: Output 'adv_file_output': Total frames output: 4
13:51:23.761: Output 'adv_file_output': Total drawn frames: 14 (15 attempted)
13:51:23.761: Output 'adv_file_output': Number of lagged frames due to rendering lag/stalls: 1 (6.7%)
13:51:23.770: ==== Recording Stop ================================================

**Comment by RytoEX** [comment]
After trying to confirm with the original reporter and not receiving a response, we're going to close this Issue.  We're fairly certain that the specific issue as originally reported was fixed back in OBS Studio 26.1.0.  If you encounter a new error related to ffmpeg-mux or `os_process_pipe_write` (e.g., "os_process_pipe_write for packet data failed"), please open up a new issue complete with a detailed description, specific reproduction steps, and an OBS log from the latest version of OBS Studio.  Issues without reproduction steps or without a log will be closed.

**Comment by dakhar48** [comment]
Thank you Ryan for caring enough to inform us of closing this thread. Really appreciate it.
